universe/Fleet.cpp

```diff
@@ -1156,6 +1156,24 @@ void Fleet::CalculateRouteTo(int target_system_id) {
     }
 }
 
+bool Fleet::Blockaded() const {
+    /** returns true if fleet is in a system and is prevented from using at least
+        one starlane entry to leave its current system */
+    std::shared_ptr<System> system = GetSystem(this->SystemID());
+
+    if (!system)
+        return false;
+
+    const std::map<int, bool> adjacent_system_IDs = system->VisibleStarlanesWormholes(this->Owner());
```

Comment:
> There's still a bug which I haven't been able to figure out: the blockade indicator seems to show for monsters and the player fleets as intended, but not for AI players. Fleet->Blockaded() seems to always return false for them. Any idea what might cause this?

I am pretty sure that part, perhaps all, of the problem is from this loop based on VisibleStarlanesWormholes() winding up empty for any empire other than your own or ALL_Empires, because VisibleStarlanesWormholes() relies on Universe::GetObjectVisibilityByEmpire, which for empireID ALL_EMPIRES will always return VIS_FULL_VISIBILITY, but for any other empire it checks visibility via the m_empire_object_visibility maps, which in the serialization process for an empire update only populates it via Universe::GetEmpireObjectVisibilityMap() with that empire's info (i.e., you don't get visibility info for other empires).

I think the solution is to not use VisibleStarlanesWormholes() at all.  The solution I would propose is to just base the loop on System::StarlanesWormholes() instead.  If you cant see the enemy fleet being blockaded along any starlane of which ___you___ are aware, then consider it not blockaded.